integrations/dynatrace/client.py

```diff
@@ -0,0 +1,104 @@
+
+    async def _get_problems(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        url = f"{self.host_url}/api/v2/problems"
+        logger.info(f"Fetching problems from {url}")
+        while True:
+            problems_response = await self.client.get(url)
+            problems_response.raise_for_status()
+            response_data = problems_response.json()
+            problems = response_data["problems"]
+            yield problems
+            next_page_key = response_data.get("nextPageKey")
+            if not next_page_key:
+                break
+            url = f"{self.host_url}/api/v2/problems?nextPageKey={next_page_key}"
+
+    async def _get_slos(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        url = f"{self.host_url}/api/v2/slo"
+        logger.info(f"Fetching slos from {url}")
+        while True:
+            slos_response = await self.client.get(url)
+            slos_response.raise_for_status()
+            response_data = slos_response.json()
+            slos = response_data["slos"]
+            yield slos
+            next_page_key = response_data.get("nextPageKey")
+            if not next_page_key:
+                break
+            url = f"{self.host_url}/api/v2/slo?nextPageKey={next_page_key}"
+
+    async def _get_entity_types(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        url = f"{self.host_url}/api/v2/entityTypes"
+        logger.info(f"Fetching entity types from {url}")
+        while True:
+            entity_types_response = await self.client.get(url)
+            entity_types_response.raise_for_status()
+            response_data = entity_types_response.json()
+            entity_types = response_data["types"]
+            yield entity_types
+            next_page_key = response_data.get("nextPageKey")
+            if not next_page_key:
+                break
+            url = f"{self.host_url}/api/v2/entityTypes?nextPageKey={next_page_key}"
+
+    async def _get_entities_from_type(
+        self, type: str
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        url = f'{self.host_url}/api/v2/?entitySelector=type("{type}")'
+        logger.info(f"Fetching entities from {url}")
+        while True:
+            entities_response = await self.client.get(url)
+            entities_response.raise_for_status()
+            response_data = entities_response.json()
+            entities = response_data["entities"]
+            yield entities
+            next_page_key = response_data.get("nextPageKey")
+            if not next_page_key:
+                break
+            url = f'{self.host_url}/api/v2/?entitySelector=type("{type}")&nextPageKey={next_page_key}'
+
+    async def _get_entities(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        async for entity_types in self._get_entity_types():
+            for entity_type in entity_types:
+                async for entities in self._get_entities_from_type(entity_type["type"]):
+                    yield entities
```

Comment:
Dynatrace has about 300 plus entity types. So instead of going through iteratively, I will suggest you run this in parallel (using asyncio) to speed up the process.

Reply:
I'm not sure we can do that easily. We won't get much performance gains. First, we can't use asyncio while running within the context of Ocean.

If I have to do this, I'll have to yield each type from the `_get_entity_types` method. This involves looping through each result in each API call. We may not have that much gain.

What do you think?